### ISSUE TITLE
[Question] How to get response body in a webview?

### ISSUE THREAD

**Issue by geeeeeeeeek** [atom/electron#4554]
I'm using a webview and wanna inspect every request in it. Here comes the question: the `details` contain everything but the **response body**. Must I retrieve the body by making another request? It seems really stupid.

```
  webview.addEventListener('did-get-response-details', (details) => {
    handleEmojiMessage(details);
  })
```

**Comment by zcbenz** [comment]
There is a Chromium issue tracking this, but not implemented yet:
https://bugs.chromium.org/p/chromium/issues/detail?id=104058

Another hack is to use the [webContents.debugger](http://electron.atom.io/docs/v0.36.8/api/web-contents/#webcontentsdebugger) API to get the body, you might get some ideas from this answer:
http://stackoverflow.com/questions/18534771/chrome-extension-how-to-get-http-response-body

Adding response body to an event, however is not a solution, it would drag the performance down when the response body is large.

**Comment by geeeeeeeeek** [comment]
@zcbenz However debugger cannot penetrate into the webview. I'm afraid this is not gonna work.

**Comment by geeeeeeeeek** [comment]
Actually there is `<webview>.openDevTools()` api provided, but we cannot retrive the debugger context.

**Comment by uinz** [comment]
@geeeeeeeeek did you resolve it?

**Comment by anatoly314** [comment]
Have somebody succeeded in this task?

**Comment by uinz** [comment]
@anatoly314 [Debugger](https://electron.atom.io/docs/api/debugger/)

I get `webContent` then use `Debugger` to handle response body

**Comment by brandonmp** [comment]
i found the snippet in [this section of the docs](https://electron.atom.io/docs/api/debugger/#class-debugger) to be more/less sufficient for this


```
const {BrowserWindow} = require('electron')
let win = new BrowserWindow()

try {
  win.webContents.debugger.attach('1.1')
} catch (err) {
  console.log('Debugger attach failed : ', err)
}

win.webContents.debugger.on('detach', (event, reason) => {
  console.log('Debugger detached due to : ', reason)
})

win.webContents.debugger.on('message', (event, method, params) => {
  if (method === 'Network.requestWillBeSent') {
    if (params.request.url === 'https://www.github.com') {
      win.webContents.debugger.detach()
    }
  }
})

win.webContents.debugger.sendCommand('Network.enable')
```

**Comment by grubshka** [comment]
I managed to handle this, for a webview. I needed to get the source only for the main document, and for all documents so I never detach the debugger, but you can easily adapt it for your needs.
The only problem I got is for the first url, at application startup : I wasn't abel to attach the debugger soon enough to get the 'Network.requestWillBeSent' event, and 'Network.getResponseBody' is not working after 'Network.responseReceived', so I ended up reloading the page.

    const webview = this.$refs.webview
    let firstShotReloaded = false

    const attachDebugger = () => {
      const debug = webview.getWebContents().debugger
      debug.attach('1.1')
      debug.on('message', (event, method, params) => {
        if (!firstShotReloaded && method === 'Network.responseReceived') {
          // XXX did not find any other way for first page load
          firstShotReloaded = true
          webview.reload()
        }
        if (method === 'Network.requestWillBeSent') {
          if (params.request.url === webview.getURL()) {
            debug.sendCommand('Network.getResponseBody', { requestId: params.requestId }, (err, data) => {
              if (err.code === undefined) {
                // XXX may check data.base64encoded boolean and decode ? Maybe not here...
                // if (data.base64encoded) ... Buffer.from(data.body, 'base64');
                this.$store.dispatch('updateStaticSource', data.body)
              }
            })
          }
        }
      })
      debug.sendCommand('Network.enable')
      webview.removeEventListener('did-start-loading', attachDebugger)
    }
    webview.addEventListener('did-start-loading', attachDebugger)

**Comment by yi** [comment]
@grubshka , how could response body be ready to get on 'Network.requestWillBeSent'?

**Comment by xdadda** [comment]
I use preload.

in html:
`<webview src="http://xxxx" preload="login.js">`
in script:
`const webview = document.querySelector('webview');
webview.addEventListener('ipc-message', (e) => { console.log(e)});`

in login.js
`var ipcRenderer = require('electron').ipcRenderer;
document.addEventListener('DOMContentLoaded' , function () {
    ipcRenderer.sendToHost('result' , document.location.href, document.body.innerText);
})`

**Comment by yi** [comment]
@xdadda I'm afraid the code you post is to get html body of the webview, not response body of http request issued by that webview. 

@grubshka's solution seems to be the right one, but there are a few issues with it which can be improved.  
  * issue 1: response body could be retrieved upon "Network.responseReceived" event. 
  * issue 2:  `debuggee.sendCommand("Network.getResponseBody", (err, res)=>{})` might end with 3002 error, which mean target request id is not available yet. If so, add some retry logic.

**Comment by xdadda** [comment]
ops, guess I misunderstood the OP question

**Comment by xgj1988** [comment]
> I managed to handle this, for a webview. I needed to get the source only for the main document, and for all documents so I never detach the debugger, but you can easily adapt it for your needs.
> The only problem I got is for the first url, at application startup : I wasn't abel to attach the debugger soon enough to get the 'Network.requestWillBeSent' event, and 'Network.getResponseBody' is not working after 'Network.responseReceived', so I ended up reloading the page.
> 
> See https://chromedevtools.github.io/devtools-protocol/tot/Network/ for all available Network debugger commands.
> 
> ```
> const webview = this.$refs.webview
> let firstShotReloaded = false
> 
> const attachDebugger = () => {
>   const debug = webview.getWebContents().debugger
>   debug.attach('1.1')
>   debug.on('message', (event, method, params) => {
>     if (!firstShotReloaded && method === 'Network.responseReceived') {
>       // XXX did not find any other way for first page load
>       firstShotReloaded = true
>       webview.reload()
>     }
>     if (method === 'Network.requestWillBeSent') {
>       if (params.request.url === webview.getURL()) {
>         debug.sendCommand('Network.getResponseBody', { requestId: params.requestId }, (err, data) => {
>           if (err.code === undefined) {
>             // XXX may check data.base64encoded boolean and decode ? Maybe not here...
>             // if (data.base64encoded) ... Buffer.from(data.body, 'base64');
>             this.$store.dispatch('updateStaticSource', data.body)
>           }
>         })
>       }
>     }
>   })
>   debug.sendCommand('Network.enable')
>   webview.removeEventListener('did-start-loading', attachDebugger)
> }
> webview.addEventListener('did-start-loading', attachDebugger)
> ```

I get the error:

 (node:8020) UnhandledPromiseRejectionWarning: Error: No data found for resource
with given identifier

**Comment by rictt** [comment]
> > I managed to handle this, for a webview. I needed to get the source only for the main document, and for all documents so I never detach the debugger, but you can easily adapt it for your needs.
> > The only problem I got is for the first url, at application startup : I wasn't abel to attach the debugger soon enough to get the 'Network.requestWillBeSent' event, and 'Network.getResponseBody' is not working after 'Network.responseReceived', so I ended up reloading the page.
> > See https://chromedevtools.github.io/devtools-protocol/tot/Network/ for all available Network debugger commands.
> > ```
> > const webview = this.$refs.webview
> > let firstShotReloaded = false
> > 
> > const attachDebugger = () => {
> >   const debug = webview.getWebContents().debugger
> >   debug.attach('1.1')
> >   debug.on('message', (event, method, params) => {
> >     if (!firstShotReloaded && method === 'Network.responseReceived') {
> >       // XXX did not find any other way for first page load
> >       firstShotReloaded = true
> >       webview.reload()
> >     }
> >     if (method === 'Network.requestWillBeSent') {
> >       if (params.request.url === webview.getURL()) {
> >         debug.sendCommand('Network.getResponseBody', { requestId: params.requestId }, (err, data) => {
> >           if (err.code === undefined) {
> >             // XXX may check data.base64encoded boolean and decode ? Maybe not here...
> >             // if (data.base64encoded) ... Buffer.from(data.body, 'base64');
> >             this.$store.dispatch('updateStaticSource', data.body)
> >           }
> >         })
> >       }
> >     }
> >   })
> >   debug.sendCommand('Network.enable')
> >   webview.removeEventListener('did-start-loading', attachDebugger)
> > }
> > webview.addEventListener('did-start-loading', attachDebugger)
> > ```
> 
> I get the error:
> 
> (node:8020) UnhandledPromiseRejectionWarning: Error: No data found for resource with given identifier

the same with you
